### PR TITLE
WEBDOM-357: Added token entity for generic replacements.

### PR DESCRIPTION
**Pull request by Jelle-S**

### CODE REVIEW

File: Service/TemplateService.php [hunk 1]
```diff
@@ -22,79 +29,112 @@ public function replaceKeys($text, array $entities = array()): string
...
+        preg_match('#\((.*?)\)#', $templateReplacementKey, $match);
+        if (isset($match[1]) && $match[1] != '') {
+            $replacementArguments = explode(',', $match[1]);
+        }
 
-                // Check if the pattern exists in our text
-                $hasMatch = preg_match('/' . $pattern . '/', $text, $matches);
+        // Complete the pattern and escape all existing special characters
+        $pattern = '[[ ' . $entityPrefix . ':' . $templateReplacementKey . ' ]]';
+        $pattern = str_replace(['(', ')', '[', ']'], ['\(', '\)', '\[', '\]'], $pattern);
```
Comment (matthijs-va) [inline Service/TemplateService.php]: Why not use preg_quote()?

File: Service/TemplateService.php [hunk 2]
```diff
@@ -22,79 +29,112 @@ public function replaceKeys($text, array $entities = array()): string
...
+        $replacePattern = $pattern;
 
-                // If we have a match for the pattern we substitute it
-                if ($hasMatch) {
-                    $hasMatches = true;
+        // Get all the arguments out of the pattern so we can match them with the real arguments
+        foreach ($replacementArguments as $replacementArgument) {
+            $pattern = str_replace($replacementArgument, '([^)]*)', $pattern);
```
Comment (matthijs-va) [inline Service/TemplateService.php]: Should "*" not be replaced with "+"?

File: Service/TemplateService.php [hunk 3]
```diff
@@ -22,79 +29,112 @@ public function replaceKeys($text, array $entities = array()): string
...
 
-        // Recursivly go trough this function until no matches are found
-        if ($hasMatches) {
-            $text = $this->replaceKeys($text, $entities);
+            $passingValue = call_user_func_array(array($passingValue, $methodName), $functionArguments);
         }
 
-        return $text;
+        // Replace the pattern with the found value
+        return preg_replace('/' . $replacePattern . '/', $passingValue, $text);
```
Comment (matthijs-va) [inline Service/TemplateService.php]: Maybe you should consider using preg_replace callback() to prevent duplicate preg_* calls.

**Comment by Jelle-S** [comment]
Looks good to me, once the tests are fixed ;)

**Pull request merged**